RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,223 @@
+        * A: Operator versioning, by design, can’t cover all torch.package use cases. So this should be out of scope.
+* **New documentation required**
+    * e2e BC-breaking guide
+        * To make a BC-breaking change update the version and write a torchscript adaptor and a mobile adaptor
+    * e2e FC-breaking guide
+        * It’s OK to add new optional keyword-only arguments as long as their default semantic preserve the operator’s current semantics
+
+Note that the proposal does not introduce an explicit version to _all_ PyTorch operators. Instead code changes are only required for updated operators with BC/FC breakage, that cannot be handled by automatic BC/FC methods. For other operators, the implicit version is v0.
+
+As an example, there’s a BC/FC breaking update on operator foo.
+
+Before:
+
+foo(Tensor self, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!)
+
+After:
+
+foo(Tensor self, **Tensor c**, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!)
+
+In schema, a Tensor argument, c is added. Note that it’s not added as a “tailing default argument”, so that BC/FC cannot be handled automatically.
+
+Accordingly, in the kernel of foo, the implementation is updated based on the new argument c. The pseudo code (it’s in Python format, but can be written in C++ as well) looks like:
+
+
+```
+def foo(Tensor self, Tensor c, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!):
+  # The original kernel implementation
+  ...
+  if not c.empty():
+    a.add_(c)
+```
+
+
+
+## Code changes (minimize the work of developer)
+
+If there is a BC/FC break with schema change in a PR, a lint error can be automatically generated, with instructions below to update the PR. 
+
+
+### Version bump
+
+Update version field in _operator_versions.yaml_.
+
+
+### BC updates
+
+The developer needs to implement a BC “upgrader” in Python. The upgrader code is put in a centralized python file, _operator_upgraders.py_. 
+
+
+
+<p id="gdcalert2" ><span style="color: red; font-weight: bold">>>>>>  gd2md-html alert: inline drawings not supported directly from Docs. You may want to copy the inline drawing to a standalone drawing and export by reference. See <a href="https://github.com/evbacher/gd2md-html/wiki/Google-Drawings-by-reference">Google Drawings by reference</a> for details. The img URL below is a placeholder. </span><br>(<a href="#">Back to top</a>)(<a href="#gdcalert3">Next alert</a>)<br><span style="color: red; font-weight: bold">>>>>> </span></p>
+
+
+![drawing](https://docs.google.com/drawings/d/12345/export/png)
+
+
+
+* Different from the upgraders defined[ here](https://github.com/pytorch/pytorch/blob/8a094e3270d2fbec6060099b7059898f4a1c104a/torch/csrc/jit/frontend/builtin_functions.cpp#L98), 
+* For some operator updates, it’s not possible to have a BC adapter. In such a case, the operator number could still help to check compatibility and to quickly detect the source of failure with meaningful error. 
```

Comment:
We should always have a BC upgrader until we get rid of them completely by switching to a BC window

Reply:
Just thinking of if there's any situation that an upgrader is not possible to be implemented for BC breaking change. Not sure how the developer experience would be affected if it happens, without make the operator pool bigger.

For FC only breaking changes there should be no upgrader.